File: src/main/python/AoC2016_04.py
#! /usr/bin/env python3
#
# Advent of Code 2016 Day 4
#

from __future__ import annotations

import re
import sys
from collections import Counter
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

REGEXP = r"([-a-z]+)-([0-9]+)\[([a-z]{5})\]$"
MATCH = r"[a-z]{9}-[a-z]{6}-[a-z]{7}$"

TEST = """\
aaaaa-bbb-z-y-x-123[abxyz]
a-b-c-d-e-f-g-h-987[abcde]
not-a-real-room-404[oarel]
totally-real-room-200[decoy]
"""


class Room(NamedTuple):
    name: str
    sector_id: int
    checksum: str

    @classmethod
    def from_input(cls, line: str) -> Room:
        m = re.search(REGEXP, line)
        assert m is not None
        return Room(m.groups()[0], int(m.groups()[1]), m.groups()[2])

    def is_real(self) -> bool:
        most_common = sorted(
            Counter(self.name.replace("-", "")).items(),
            key=lambda item: item[1] * -100 + ord(item[0]),
        )
        return "".join(x for x, _ in most_common[:5]) == self.checksum

    def decrypt(self) -> str:
        shift = self.sector_id % 26
        return "".join(
            chr((ord(c) + shift - 97) % 26 + 97) if c != "-" else " "
            for c in self.name
        )


Input = list[Room]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [Room.from_input(line) for line in input_data]

    def part_1(self, rooms: Input) -> int:
        return sum(room.sector_id for room in rooms if room.is_real())

    def part_2(self, rooms: Input) -> int:
        return next(
            room.sector_id
            for room in rooms
            if re.match(MATCH, room.name) is not None
            and room.decrypt() == "northpole object storage"
        )

    @aoc_samples((("part_1", TEST, 1514),))
    def samples(self) -> None:
        pass


solution = Solution(2016, 4)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2016_04.java
import static com.github.pareronia.aoc.AssertUtils.assertTrue;
import static com.github.pareronia.aoc.Utils.toAString;
import static java.util.Comparator.comparing;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.github.pareronia.aoc.Counter;
import com.github.pareronia.aoc.Counter.Entry;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2016_04
        extends SolutionBase<List<AoC2016_04.Room>, Integer, Integer> {
    
    private static final Pattern REGEXP = Pattern.compile("([-a-z]+)-([0-9]+)\\[([a-z]{5})\\]$");
    private static final Pattern MATCH = Pattern.compile("[a-z]{9}-[a-z]{6}-[a-z]{7}$");
    
	private AoC2016_04(final boolean debug) {
		super(debug);
	}
	
	public static final AoC2016_04 create() {
		return new AoC2016_04(false);
	}

	public static final AoC2016_04 createDebug() {
		return new AoC2016_04(true);
	}

	@Override
    protected List<Room> parseInput(final List<String> inputs) {
        return inputs.stream().map(Room::fromInput).toList();
    }

    @Override
	public Integer solvePart1(final List<Room> rooms) {
		return rooms.stream()
		        .filter(Room::isReal)
		        .mapToInt(Room::sectorId)
		        .sum();
	}

	@Override
	public Integer solvePart2(final List<Room> rooms) {
	    return rooms.stream()
	            .filter(r -> MATCH.matcher(r.name()).matches())
	            .filter(r -> r.decrypt().equals("northpole object storage"))
	            .map(Room::sectorId)
	            .findFirst().orElseThrow();
	}

	@Samples({
	    @Sample(method = "part1", input = TEST, expected = "1514")
	})
	public static void main(final String[] args) throws Exception {
	    AoC2016_04.create().run();
	}

	private static final String TEST = """
	        aaaaa-bbb-z-y-x-123[abxyz]
	        a-b-c-d-e-f-g-h-987[abcde]
	        not-a-real-room-404[oarel]
	        totally-real-room-200[decoy]
	        """;
	
	record Room(String name, int sectorId, String checkum) {
	    
	    public static Room fromInput(final String line) {
	        final Matcher m = REGEXP.matcher(line);
            assertTrue(m.matches(), () -> "Expected match");
            return new Room(m.group(1), Integer.parseInt(m.group(2)), m.group(3));
	    }
	    
	    public boolean isReal() {
	        return new Counter<>(
	                Utils.asCharacterStream(this.name.replace("-", "")))
	            .mostCommon().stream()
	            .sorted(comparing(e -> e.count() * -100 + e.value().charValue()))
	            .limit(5)
	            .map(Entry::value)
	            .collect(toAString())
	            .equals(this.checkum);
	    }
	    
	    private Character decrypt(final Character c, final int shift) {
	        if (c == '-') {
	            return ' ';
	        }
	        return StringOps.nextLetter(c, shift);
	    }
	    
	    public String decrypt() {
	        final int shift = this.sectorId % 26;
	        return Utils.asCharacterStream(this.name)
	                .map(c -> decrypt(c, shift))
	                .collect(toAString());
	    }
	}
}
